clients/clearly-defined/src/main/kotlin/ClearlyDefinedService.kt

```diff
@@ -289,6 +289,28 @@ interface ClearlyDefinedService {
     ): ResponseBody
 }
 
+suspend fun ClearlyDefinedService.getCuration(coordinates: Coordinates): Curation =
+    @Suppress("DestructuringDeclarationWithTooManyEntries")
+    coordinates.strings.let { (_, _, namespace, name, revision) ->
+        getCuration(coordinates.type, coordinates.provider, namespace, name, revision)
+    }
+
+suspend fun ClearlyDefinedService.harvestTools(coordinates: Coordinates): List<String> =
+    @Suppress("DestructuringDeclarationWithTooManyEntries")
```

Comment:
The called functions take `type` and `provider` as types enums, not as strings, so their string representations are "discarded" and their typed values are used instead. I liked that more than reconstructing the enums from the strings.

Reply:
Hm, this means the first two items in the `strings` list are unused.

Do I get it correctly, that the only reason why the list is introduced is for reusing
`revision.orEmpty()` and ensuring that `-` represents the empty namespace.

I wonder if alternatively making `revision` not null in Coordinates, and disallowing empty strings as namespace, enforcing `-`, then all properties could just be as is - which would make the `string` property unnecessary. This code the could just look like

`with(coordinates) { harvestTools(type, provider, namespace, name, revision) }`

What do you think?

Reply:
> Do I get it correctly, that the only reason why the list is introduced is for reusing
> `revision.orEmpty()` and ensuring that `-` represents the empty namespace.

Basically yes.

> What do you think?

The `Coordinates` class is also used in API responses, so we cannot make `revision` (or `namespace`) non-null as then deserialization of results would fail.